### PR TITLE
changed a few data types to auto, minor bug fixes and examples

### PR DESCRIPTION
**Pull request by cs-mshah**

### CODE REVIEW

File: templates/ordered_set.cpp [hunk 1]
```diff
@@ -45,23 +48,64 @@ struct ordered_multiset {
...
 	num_t find_by_order(int k) { return (*vals.find_by_order(k)).first; }
 	int order_of_key(num_t k) { return vals.order_of_key(make_pair(k - 1, 0)); }
-	pair<num_t, int>* begin() { return vals.begin(); }
-	pair<num_t, int>* end() { return vals.end(); }
-	pair<num_t, int>* rbegin() { return vals.rbegin(); }
-	pair<num_t, int>* rend() { return vals.rend(); }
+	auto begin() { return vals.begin(); }
+	auto end() { return vals.end(); }
+	auto rbegin() { return vals.rbegin(); }
+	auto rend() { return vals.rend(); }
```
Comment (cs-mshah) [inline templates/ordered_set.cpp:54]: the data type was causing some error while compiling, so just changed it to auto, and now iteration becomes easy with `for(auto x: oms)`.

**Comment by galencolin** [comment]
Cool, I didn't even know the auto stuff was possible

File: templates/ordered_set.cpp [hunk 2]
```diff
@@ -45,23 +48,64 @@ struct ordered_multiset {
...
 	int size() { return vals.size(); }
 	void clear() { vals.clear(); best.clear(); }
-	int count(num_t k) { return vals.order_of_key(k + 1) - vals.order_of_key(k); }
```
Comment (cs-mshah) [inline templates/ordered_set.cpp:54]: needs to be of type `pair<num_t, int>` in vals.order_of_key()

File: templates/ordered_set.cpp [hunk 3]
```diff
@@ -45,23 +48,64 @@ struct ordered_multiset {
...
+	for(auto m:oms) // iterate over ordered multiset
+	{
+		cout<<m.first<<" ";
+	}
+	cout<<endl;
+	cout<<oms.find_by_order(2)<<endl;
+	cout<<oms.order_of_key(26)<<endl;
+	cout<<oms.count(24)<<endl;
+	cout<<oms.lower_bound(24)->first<<endl; // use in this way
+	cout<<oms.upper_bound(23)->first<<endl;
```
Comment (cs-mshah) [inline templates/ordered_set.cpp:98]: examples demonstrating the usage.

**Pull request merged**